statistics/handle/cache/statscacheinner.go

```diff
@@ -26,64 +26,94 @@ import (
-	"github.com/pingcap/tidb/statistics/handle/internal/cache/mapcache"
+	"github.com/pingcap/tidb/statistics/handle/cache/internal"
+	"github.com/pingcap/tidb/statistics/handle/cache/internal/lru"
+	"github.com/pingcap/tidb/statistics/handle/cache/internal/mapcache"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/sqlexec"
 	"github.com/pingcap/tidb/util/syncutil"
 )
 
-func newStatsCache() statsCache {
+type TableStatsOption struct {
+	byQuery bool
+}
+
+func (t *TableStatsOption) ByQuery() bool {
+	return t.byQuery
+}
+
+// TableStatsOpt used to edit getTableStatsOption
+type TableStatsOpt func(*TableStatsOption)
+
+// WithTableStatsByQuery indicates user needed
+func WithTableStatsByQuery() TableStatsOpt {
+	return func(option *TableStatsOption) {
+		option.byQuery = true
+	}
+}
+
+func NewStatsCacheWrapper() StatsCacheWrapper {
 	enableQuota := config.GetGlobalConfig().Performance.EnableStatsCacheMemQuota
 	if enableQuota {
 		capacity := variable.StatsCacheMemQuota.Load()
-		return statsCache{
+		return StatsCacheWrapper{
 			StatsCacheInner: lru.NewStatsLruCache(capacity),
 		}
 	}
-	return statsCache{
+	return StatsCacheWrapper{
 		StatsCacheInner: mapcache.NewMapCache(),
 	}
 }
 
-// statsCache caches the tables in memory for Handle.
-type statsCache struct {
-	cache.StatsCacheInner
+// StatsCacheWrapper caches the tables in memory for Handle.
+type StatsCacheWrapper struct {
+	internal.StatsCacheInner
 	// version is the latest version of cache. It is bumped when new records of `mysql.stats_meta` are loaded into cache.
 	version uint64
 	// minorVersion is to differentiate the cache when the version is unchanged while the cache contents are
 	// modified indeed. This can happen when we load extra column histograms into cache, or when we modify the cache with
 	// statistics feedbacks, etc. We cannot bump the version then because no new changes of `mysql.stats_meta` are loaded,
-	// while the override of statsCache is in a copy-on-write way, to make sure the statsCache is unchanged by others during the
+	// while the override of StatsCacheWrapper is in a copy-on-write way, to make sure the StatsCacheWrapper is unchanged by others during the
 	// the interval of 'copy' and 'write', every 'write' should bump / check this minorVersion if the version keeps
 	// unchanged.
-	// This bump / check logic is encapsulated in `statsCache.update` and `updateStatsCache`, callers don't need to care
+	// This bump / check logic is encapsulated in `StatsCacheWrapper.update` and `updateStatsCache`, callers don't need to care
 	// about this minorVersion actually.
 	minorVersion uint64
 }
 
-func (sc statsCache) len() int {
+// Len returns the number of tables in the cache.
+func (sc StatsCacheWrapper) Len() int {
 	return sc.StatsCacheInner.Len()
 }
 
-func (sc statsCache) copy() statsCache {
-	newCache := statsCache{
+// Copy copies the cache.
+func (sc StatsCacheWrapper) Copy() StatsCacheWrapper {
```

Comment:
Looks like we should return the pointer here.

Reply:
Sure, I think it is a bad code. but It will be refactored as soon as possible. It is unnecessary for it to copy all the data every time.